src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -497,6 +494,25 @@ Pipe Kafka::read(
     return pipe;
 }
 
+std::shared_ptr<RdKafka::Consumer> Kafka::getConsumer()
+{
+    auto new_consumer = std::make_shared<RdKafka::Consumer>(*conf, poll_timeout_ms, getLoggerName());
+    std::weak_ptr<RdKafka::Consumer> ref = new_consumer;
+
+    std::lock_guard<std::mutex> lock(consumer_mutex);
+    for (auto & consumer : consumers)
+    {
+        /// if a consumer is gone, replace it with the new one, so that `consumers` won't keep growing forever.
+        if (consumer.expired())
+        {
+            consumer.swap(ref);
+            return new_consumer;
+        }
+    }
+    consumers.push_back(ref);
```

Comment:
Yes, in theory that's correct. However, ad-hoc queries will eventually stop at some point, only consumers used in MVs will be long-live. So, in practical, the vector won't keep growing forever. Also, we have already had a setting that limits the number of queries can running in proton at the same time, so, I don't feel like that we need another limit on the number of consumers.

Please let me know if you have different ideas.

Reply:
this dependency / assumption seems too deep. I would like to make sure every component has bounded queue since we have seen things goes wild when the assumption breaks. Also we may not like to creating 1000 consumers when the ad-hoc concurrent query is 1000 but this is a different concern.

Reply:
Added a limit.